### PR TITLE
Add projects API

### PR DESCRIPTION
**Pull request by jfreda**
This PR adds APIs for creating and updating projects and their associated related resources (both API versions `v1` and `v2` are handled by the same code).

### New API Endpoints

- `/api/v1/projects`: `GET`, `POST`
- `/api/v1/projects/{project_id}`: `GET`, `PATCH`
- `/api/v1/projects/{project_id}/related-resources`: `GET`, `POST`, `PUT`
- `/api/v2/projects`: `GET`, `POST`
- `/api/v2/projects/{project_id}`: `GET`, `PATCH`
- `/api/v2/projects/{project_id}/related-resources`: `GET`, `POST`, `PUT`

### CODE REVIEW

File: internal/api/v2/projects.go [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+						srv.Logger.Warn("project not found", logArgs...)
+						http.Error(w, "Project not found", http.StatusNotFound)
+						return
+					} else {
+						srv.Logger.Error("error getting project from database",
+							append([]interface{}{
+								"error", err,
+							}, logArgs...)...)
+						http.Error(
+							w, "Error processing requst", http.StatusInternalServerError)
```
Comment (anubhavmishra) [inline internal/api/v2/projects.go:284]:
```suggestion
							w, "Error processing request", http.StatusInternalServerError)
```

File: internal/api/v2/projects_related_resources.go [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+				srv.Logger.Warn("project not found", logArgs...)
+				http.Error(w, "Project not found", http.StatusNotFound)
+				return
+			} else {
+				srv.Logger.Error("error getting project from database",
+					append([]interface{}{
+						"error", err,
+					}, logArgs...)...)
+				http.Error(
+					w, "Error processing requst", http.StatusInternalServerError)
```
Comment (anubhavmishra) [inline internal/api/v2/projects_related_resources.go:77]:
```suggestion
					w, "Error processing request", http.StatusInternalServerError)
```

File: internal/api/v2/projects_related_resources.go [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+		// Get typed related resources.
+		elrrs, hdrrs, err := proj.GetRelatedResources(srv.DB)
+		if err != nil {
+			srv.Logger.Error("error getting related resources",
+				append([]interface{}{
+					"error", err,
+				}, logArgs...)...)
+			http.Error(
+				w, "Error processing requst", http.StatusInternalServerError)
```
Comment (anubhavmishra) [inline internal/api/v2/projects_related_resources.go:90]:
```suggestion
				w, "Error processing request", http.StatusInternalServerError)
```

File: internal/api/v2/projects_related_resources.go [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+		// Add external link related resources.
+		for _, elrr := range elrrs {
+			if err := elrr.Get(srv.DB); err != nil {
+				srv.Logger.Error(
+					"error getting external link related resource from database",
+					append([]interface{}{
+						"error", err,
+					}, logArgs...)...)
+				http.Error(
+					w, "Error processing requst", http.StatusInternalServerError)
```
Comment (anubhavmishra) [inline internal/api/v2/projects_related_resources.go:108]:
```suggestion
					w, "Error processing request", http.StatusInternalServerError)
```

File: internal/api/v2/projects_related_resources.go [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+			model := models.Document{
+				GoogleFileID: hdrr.Document.GoogleFileID,
+			}
+			if err := model.Get(srv.DB); err != nil {
+				srv.Logger.Error("error getting document from database",
+					append([]interface{}{
+						"error", err,
+					}, logArgs...)...)
+				http.Error(
+					w, "Error processing requst", http.StatusInternalServerError)
```
Comment (anubhavmishra) [inline internal/api/v2/projects_related_resources.go:133]:
```suggestion
					w, "Error processing request", http.StatusInternalServerError)
```

File: internal/api/v2/projects_related_resources.go [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+				Document: models.Document{
+					GoogleFileID: hdrr.Document.GoogleFileID,
+				},
+			}); err != nil {
+				srv.Logger.Error("error getting reviews for document",
+					append([]interface{}{
+						"error", err,
+					}, logArgs...)...)
+				http.Error(
+					w, "Error processing requst", http.StatusInternalServerError)
```
Comment (anubhavmishra) [inline internal/api/v2/projects_related_resources.go:149]:
```suggestion
					w, "Error processing request", http.StatusInternalServerError)
```

File: internal/api/v2/projects_related_resources.go [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+			// Convert database model to a document.
+			doc, err := document.NewFromDatabaseModel(
+				model, reviews)
+			if err != nil {
+				srv.Logger.Error("error converting database model to document type",
+					append([]interface{}{
+						"error", err,
+					}, logArgs...)...)
+				http.Error(
+					w, "Error processing requst", http.StatusInternalServerError)
```
Comment (anubhavmishra) [inline internal/api/v2/projects_related_resources.go:162]:
```suggestion
					w, "Error processing request", http.StatusInternalServerError)
```

File: internal/api/v2/projects_related_resources.go [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+		w.WriteHeader(http.StatusOK)
+		enc := json.NewEncoder(w)
+		err = enc.Encode(resp)
+		if err != nil {
+			srv.Logger.Error("error encoding response",
+				append([]interface{}{
+					"error", err,
+				}, logArgs...)...)
+			http.Error(
+				w, "Error processing requst", http.StatusInternalServerError)
```
Comment (anubhavmishra) [inline internal/api/v2/projects_related_resources.go:188]:
```suggestion
				w, "Error processing request", http.StatusInternalServerError)
```

File: internal/api/v2/projects_related_resources.go [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+				srv.Logger.Warn("project not found", logArgs...)
+				http.Error(w, "Project not found", http.StatusNotFound)
+				return
+			} else {
+				srv.Logger.Error("error getting project from database",
+					append([]interface{}{
+						"error", err,
+					}, logArgs...)...)
+				http.Error(
+					w, "Error processing requst", http.StatusInternalServerError)
```
Comment (anubhavmishra) [inline internal/api/v2/projects_related_resources.go:222]:
```suggestion
					w, "Error processing request", http.StatusInternalServerError)
```

File: internal/api/v2/projects_related_resources.go [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+		}
+
+		// Replace related resources for project.
+		if err := proj.ReplaceRelatedResources(srv.DB, elrrs, hdrrs); err != nil {
+			srv.Logger.Error("error replacing related resources for document",
+				append([]interface{}{
+					"error", err,
+				}, logArgs...)...)
+			http.Error(
+				w, "Error processing requst", http.StatusInternalServerError)
```
Comment (anubhavmishra) [inline internal/api/v2/projects_related_resources.go:261]:
```suggestion
				w, "Error processing request", http.StatusInternalServerError)
```

File: internal/cmd/commands/server/server.go [hunk 11]
```diff
@@ -342,6 +342,8 @@ func (c *Command) Run(args []string) int {
 			api.MeSubscriptionsHandler(cfg, c.Log, goog, db)},
 		{"/api/v1/people", api.PeopleDataHandler(cfg, c.Log, goog)},
 		{"/api/v1/products", api.ProductsHandler(cfg, algoSearch, c.Log)},
+		{"/api/v1/products", apiv2.ProductsHandler(srv)},
+		{"/api/v1/projects", apiv2.ProjectsHandler(srv)},
```
Comment (anubhavmishra) [inline internal/cmd/commands/server/server.go:346]: Not sure why we would add `v1` endpoints for projects given it is a brand new feature? Is this to support some type of backwards incompatibility?

**Review by anubhavmishra — CHANGES_REQUESTED** [review]

File: internal/api/v2/projects_related_resources.go (hunk 2 above)
Comment (jfreda) [reply internal/api/v2/projects_related_resources.go:77]: Good catch, I really leaned into this typo haha 🙈

File: internal/cmd/commands/server/server.go (hunk 11 above)
Comment (jfreda) [reply internal/cmd/commands/server/server.go:346]: The API version is (more or less, currently, but there's a draft PR #399 that might make this explicit) global for all frontend `fetch` calls, so I wanted to make sure that this is supported in v1 of the API as well (while it exists). The same code is used for both API versions.

**Review by anubhavmishra — APPROVED** [review]

File: internal/cmd/commands/server/server.go (hunk 11 above)
Comment (anubhavmishra) [reply internal/cmd/commands/server/server.go:346]: Makes sense!

**Pull request merged**